homeassistant/components/agent_dvr/__init__.py

```diff
@@ -0,0 +1,129 @@
+"""Support for Agent."""
+import logging
+
+from agent.a import Agent
+
+from homeassistant.auth.permissions.const import POLICY_CONTROL
+from homeassistant.const import ATTR_ENTITY_ID, ENTITY_MATCH_ALL
+from homeassistant.exceptions import Unauthorized, UnknownUser
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_call_later
+from homeassistant.helpers.service import async_extract_entity_ids
+
+from .camera import CAMERA_SERVICES
+from .const import DEFAULT_RETRY, DOMAIN as AGENT_DOMAIN, SERVER_URL
+from .helpers import service_signal
+
+ATTRIBUTION = "ispyconnect.com"
+DEFAULT_BRAND = "Agent DVR by ispyconnect.com"
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_TIMEOUT = 10
+FORWARDS = ["camera"]
+
+
+async def async_setup(hass, config):
+    """Old way to set up integrations."""
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up the Agent component."""
+
+    async def retry_setup(now):
+        """Retry setup if a connection/timeout happens on Agent API."""
+        await async_setup_entry(hass, config_entry)
+
+    hass.data.setdefault(AGENT_DOMAIN, {})
+
+    server_origin = config_entry.data[SERVER_URL]
+    if not server_origin:
+        return False
+    agent_client = Agent(server_origin)
+    await agent_client.update()
+    if not agent_client.is_available:
+        async_call_later(hass, DEFAULT_RETRY, retry_setup)
+        return True
+
+    agent_client.deviceList = await agent_client.get_devices()
+
+    hass.data[AGENT_DOMAIN][server_origin] = agent_client
+
+    device_registry = await dr.async_get_registry(hass)
+
+    device_registry.async_get_or_create(
+        config_entry_id=config_entry.entry_id,
+        identifiers={(AGENT_DOMAIN, agent_client.unique)},
+        manufacturer="iSpyConnect",
+        name="Agent " + agent_client.name,
+        model="Agent DVR",
+        sw_version=agent_client.version,
+    )
+
+    for forward in FORWARDS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, forward)
+        )
+
+    def have_permission(user, entity_id):
+        return not user or user.permissions.check_entity(entity_id, POLICY_CONTROL)
+
+    async def async_extract_from_service(call):
```

Comment:
You don't need all this to implement entity services, see: 

<https://developers.home-assistant.io/docs/dev_101_services#entity-services>